pkg/common/log/reopen.go

```diff
@@ -0,0 +1,106 @@
+package log
+
+import (
+	"context"
+	"io"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+)
+
+const (
+	_fileFlags    = os.O_APPEND | os.O_CREATE | os.O_WRONLY
+	_fileMode     = 0640
+	_reopenSignal = syscall.SIGUSR2
+)
+
+var _ ReopenableWriteCloser = (*ReopenableFile)(nil)
+
+// Reopener inspired by https://github.com/client9/reopen
+type Reopener interface {
+	Reopen() error
+}
+
+type ReopenableWriteCloser interface {
+	Reopener
+	io.WriteCloser
+}
+
+type ReopenableFile struct {
+	name string
+	f    *os.File
+	mu   sync.Mutex
+}
+
+func NewReopenableFile(name string) (*ReopenableFile, error) {
+	file, err := os.OpenFile(name, _fileFlags, _fileMode)
+	if err != nil {
+		return nil, err
+	}
+	return &ReopenableFile{
+		name: name,
+		f:    file,
+	}, nil
+}
+
+func (r *ReopenableFile) Reopen() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.f != nil {
+		if err := r.f.Close(); err != nil {
+			return err
+		}
+		r.f = nil
+	}
+	newFile, err := os.OpenFile(r.name, _fileFlags, _fileMode)
+	if err != nil {
+		r.f = nil
+		return err
+	}
+	r.f = newFile
+	return nil
+}
+
+func (r *ReopenableFile) Write(p []byte) (n int, err error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	return r.f.Write(p)
+}
+
+func (r *ReopenableFile) Close() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	return r.f.Close()
+}
+
+// Name implements part of os.FileInfo without needing a lock on the
+// underlying file.
+func (r *ReopenableFile) Name() string {
+	return r.name
+}
```

Comment:
Maybe I should add a `Stat()` method which locks & returns `r.f.Stat()`?

Reply:
> I don't think we really need it since logrus accepts a generic io.Writer for its output stream. I'm not seeing a use case for Name() either other than for tests. The test is already reaching into the f *os.File field for Stat() though, so I think it may be cleaner to just rely directly on r.f for Name() as well.

I missed that detail. Agreed. I only added this to satisfy that test to get the PR up.